### PR TITLE
TMA-1081: fail UFB when multiple_project mode column is missing from UB input

### PR DESCRIPTION
**Pull request by kubamahnert**

### CODE REVIEW

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/609/ : SUCCESS in 11s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1431/ : FAILURE in 2m 42s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1700/ : SUCCESS in 3m 05s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/611/ : SUCCESS in 56s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1434/ : FAILURE in 1m 05s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1703/ : SUCCESS in 2m 32s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/615/ : SUCCESS in 7s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1437/ : FAILURE in 1m 13s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1705/ : SUCCESS in 1m 11s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/637/ : SUCCESS in 42s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1466/ : FAILURE in 1m 39s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1727/ : FAILURE in 1m 47s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/643/ : SUCCESS in 6s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1473/ : FAILURE in 3m 18s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1734/ : SUCCESS in 1m 24s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/647/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1475/ : FAILURE in 1m 05s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1736/ : FAILURE in 1m 07s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/648/ : SUCCESS in 11s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1476/ : FAILURE in 1m 31s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1737/ : FAILURE in 1m 35s

File: lib/gooddata/lcm/actions/collect_users_brick_users.rb [hunk 1]
```diff
@@ -39,12 +48,14 @@ def call(params)
                       headers: true,
                       return_headers: false,
                       encoding: 'utf-8') do |row|
+            pid = row[params.multiple_projects_column]
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/collect_users_brick_users.rb]: info: Calls 'params.multiple_projects_column' 2 times (DuplicateMethodCall) (reek)

File: spec/lcm/integration/user_provisioning_end_to_end_spec.rb [hunk 2]
```diff
@@ -6,56 +6,81 @@
...
   blacklists = $generated_uniq_labels_with_value || {}
   blacklist = blacklists[project.pid] || []
 
-  atts = project.attributes.to_a
+  atts = project.attributes.to_a.sort_by(&:identifier)
   attribute = atts.shift while !attribute || attribute.date_attribute? || !attribute.primary_label || (blacklist.include? attribute)
   label = attribute.primary_label
-  value = label.values.find { |v| !v[:value].empty? }
+  sorted_values = label.values.to_a.sort { |x, y| x[:value] <=> y[:value] }
+  notempty = sorted_values.select { |v| !v[:value].empty? }
```
Comment (yenkins) [inline spec/lcm/integration/user_provisioning_end_to_end_spec.rb]: info: Calls 'v[:value]' 2 times (DuplicateMethodCall) (reek)

Comment (yenkins) [inline spec/lcm/integration/user_provisioning_end_to_end_spec.rb]: info: Has the variable name 'v' (UncommunicativeVariableName) (reek)

File: spec/lcm/integration/user_provisioning_end_to_end_spec.rb [hunk 3]
```diff
@@ -6,56 +6,81 @@
...
+end
 
-    @projects.each do |p|
-      filters = p.user_filters.to_a
-      expect(filters.length).to eq(@mufs.count / @projects.count)
-      @mufs.each do |u|
-        expect(filters.find { |f| f.related.email == u[:login] })
-      end
+def test_user_filters_brick(options = {})
+  test_context = options[:test_context]
```
Comment (yenkins) [inline spec/lcm/integration/user_provisioning_end_to_end_spec.rb]: info: Refers to 'options' more than self (maybe move it to another class?) (FeatureEnvy) (reek)

File: spec/lcm/integration/user_provisioning_end_to_end_spec.rb [hunk 4]
```diff
@@ -6,56 +6,81 @@
...
+      expect(filters.find { |f| f.related.email == u[:login] })
     end
   end
 end
 
-shared_context 'users brick' do
-  it 'updates users' do
-    BrickRunner.users_brick context: @test_context, template_path: '../params/users_brick_e2e.json.erb'
+def test_users_brick(options = {})
+  test_context = options[:test_context]
```
Comment (yenkins) [inline spec/lcm/integration/user_provisioning_end_to_end_spec.rb]: info: Refers to 'options' more than self (maybe move it to another class?) (FeatureEnvy) (reek)

File: spec/lcm/integration/user_provisioning_end_to_end_spec.rb [hunk 5]
```diff
@@ -6,56 +6,81 @@
...
 end
 
-shared_context 'user filters brick' do
-  it 'updates user filters' do
-    BrickRunner.user_filters_brick context: @test_context, template_path: '../params/user_filters_brick_e2e.json.erb'
+def set_up_random_muf_data(project, user_data, complexity = 2)
+  label, values = uniq_label_with_value project, complexity
+  puts "Settings up MUF referencing value #{values} of attribute #{label.title}"
+  mufs = user_data.map do |u|
+    values.map do |v|
```
Comment (yenkins) [inline spec/lcm/integration/user_provisioning_end_to_end_spec.rb]: info: Contains iterators nested 2 deep (NestedIterators) (reek)

Comment (yenkins) [inline spec/lcm/integration/user_provisioning_end_to_end_spec.rb]: info: Has the variable name 'v' (UncommunicativeVariableName) (reek)

File: spec/lcm/integration/user_provisioning_end_to_end_spec.rb [hunk 6]
```diff
@@ -6,56 +6,81 @@
...
+  projects = options[:projects]
+  muf_complexity = options[:muf_complexity] || 2
+
+  BrickRunner.user_filters_brick context: test_context, template_path: '../params/user_filters_brick_e2e.json.erb'
+
+  projects.each do |p|
+    filters = p.user_filters.to_a
+    expect(filters.length).to eq(mufs.count / projects.count / muf_complexity)
+    mufs.each do |u|
+      expect(filters.find { |f| f.related.email == u[:login] })
```
Comment (yenkins) [inline spec/lcm/integration/user_provisioning_end_to_end_spec.rb]: info: Contains iterators nested 3 deep (NestedIterators) (reek)

Comment (yenkins) [inline spec/lcm/integration/user_provisioning_end_to_end_spec.rb]: info: Has the variable name 'f' (UncommunicativeVariableName) (reek)

File: spec/lcm/integration/user_provisioning_end_to_end_spec.rb [hunk 7]
```diff
@@ -6,56 +6,81 @@
...
   blacklists[project.pid] = blacklist.push attribute
   $generated_uniq_labels_with_value = blacklists
-  [label, value[:value]]
+  [label, values.map { |v| v[:value] }]
 end
 
-shared_context 'user filters brick' do
-  it 'updates user filters' do
-    BrickRunner.user_filters_brick context: @test_context, template_path: '../params/user_filters_brick_e2e.json.erb'
+def set_up_random_muf_data(project, user_data, complexity = 2)
```
Comment (yenkins) [inline spec/lcm/integration/user_provisioning_end_to_end_spec.rb]: info: Has approx 7 statements (TooManyStatements) (reek)

File: spec/lcm/integration/user_provisioning_end_to_end_spec.rb [hunk 8]
```diff
@@ -6,56 +6,81 @@
...
+  mufs
+end
 
-    @projects.each do |p|
-      filters = p.user_filters.to_a
-      expect(filters.length).to eq(@mufs.count / @projects.count)
-      @mufs.each do |u|
-        expect(filters.find { |f| f.related.email == u[:login] })
-      end
+def test_user_filters_brick(options = {})
```
Comment (yenkins) [inline spec/lcm/integration/user_provisioning_end_to_end_spec.rb]: info: Has approx 11 statements (TooManyStatements) (reek)

File: spec/lcm/integration/user_provisioning_end_to_end_spec.rb [hunk 9]
```diff
@@ -6,56 +6,81 @@
...
+    mufs.each do |u|
+      expect(filters.find { |f| f.related.email == u[:login] })
     end
   end
 end
 
-shared_context 'users brick' do
-  it 'updates users' do
-    BrickRunner.users_brick context: @test_context, template_path: '../params/users_brick_e2e.json.erb'
+def test_users_brick(options = {})
```
Comment (yenkins) [inline spec/lcm/integration/user_provisioning_end_to_end_spec.rb]: info: Has approx 7 statements (TooManyStatements) (reek)

File: spec/lcm/integration/user_provisioning_end_to_end_spec.rb [hunk 10]
```diff
@@ -6,56 +6,81 @@
 require_relative 'support/s3_helper'
 require_relative 'brick_runner'
 
-def uniq_label_with_value(project)
+def uniq_label_with_value(project, value_count)
```
Comment (yenkins) [inline spec/lcm/integration/user_provisioning_end_to_end_spec.rb]: info: Has approx 14 statements (TooManyStatements) (reek)

Comment (yenkins) [inline spec/lcm/integration/user_provisioning_end_to_end_spec.rb]: info: Doesn't depend on instance state (maybe move it to another class?) (UtilityFunction) (reek)

File: spec/lcm/integration/user_provisioning_end_to_end_spec.rb [hunk 11]
```diff
@@ -6,56 +6,81 @@
...
+  [label, values.map { |v| v[:value] }]
 end
 
-shared_context 'user filters brick' do
-  it 'updates user filters' do
-    BrickRunner.user_filters_brick context: @test_context, template_path: '../params/user_filters_brick_e2e.json.erb'
+def set_up_random_muf_data(project, user_data, complexity = 2)
+  label, values = uniq_label_with_value project, complexity
+  puts "Settings up MUF referencing value #{values} of attribute #{label.title}"
+  mufs = user_data.map do |u|
```
Comment (yenkins) [inline spec/lcm/integration/user_provisioning_end_to_end_spec.rb]: info: Has the variable name 'u' (UncommunicativeVariableName) (reek)

File: spec/lcm/integration/user_provisioning_end_to_end_spec.rb [hunk 12]
```diff
@@ -6,56 +6,81 @@
...
-      end
+def test_user_filters_brick(options = {})
+  test_context = options[:test_context]
+  mufs = options[:mufs]
+  projects = options[:projects]
+  muf_complexity = options[:muf_complexity] || 2
+
+  BrickRunner.user_filters_brick context: test_context, template_path: '../params/user_filters_brick_e2e.json.erb'
+
+  projects.each do |p|
```
Comment (yenkins) [inline spec/lcm/integration/user_provisioning_end_to_end_spec.rb]: info: Has the variable name 'p' (UncommunicativeVariableName) (reek)

File: spec/lcm/integration/user_provisioning_end_to_end_spec.rb [hunk 13]
```diff
@@ -6,56 +6,81 @@
...
+    filters = p.user_filters.to_a
+    expect(filters.length).to eq(mufs.count / projects.count / muf_complexity)
+    mufs.each do |u|
```
Comment (yenkins) [inline spec/lcm/integration/user_provisioning_end_to_end_spec.rb]: info: Has the variable name 'u' (UncommunicativeVariableName) (reek)

File: spec/lcm/integration/user_provisioning_end_to_end_spec.rb [hunk 14]
```diff
@@ -6,56 +6,81 @@
...
+  user_data = options[:user_data]
 
-    @projects.each do |p|
-      users = p.users.to_a
-      expect(users.length).to eq((@user_data.count / @projects.count) + 1) # rubydev+admin is in the projects as well (as he is the admin)
-      # TODO: check the data is the same
-    end
+  BrickRunner.users_brick context: test_context, template_path: '../params/users_brick_e2e.json.erb'
+
+  projects.each do |p|
```
Comment (yenkins) [inline spec/lcm/integration/user_provisioning_end_to_end_spec.rb]: info: Has the variable name 'p' (UncommunicativeVariableName) (reek)

File: spec/lcm/integration/user_provisioning_end_to_end_spec.rb [hunk 15]
```diff
@@ -6,56 +6,81 @@
...
+def uniq_label_with_value(project, value_count)
   blacklists = $generated_uniq_labels_with_value || {}
   blacklist = blacklists[project.pid] || []
 
-  atts = project.attributes.to_a
+  atts = project.attributes.to_a.sort_by(&:identifier)
   attribute = atts.shift while !attribute || attribute.date_attribute? || !attribute.primary_label || (blacklist.include? attribute)
   label = attribute.primary_label
-  value = label.values.find { |v| !v[:value].empty? }
+  sorted_values = label.values.to_a.sort { |x, y| x[:value] <=> y[:value] }
```
Comment (yenkins) [inline spec/lcm/integration/user_provisioning_end_to_end_spec.rb]: info: Has the variable name 'x' (UncommunicativeVariableName) (reek)

Comment (yenkins) [inline spec/lcm/integration/user_provisioning_end_to_end_spec.rb]: info: Has the variable name 'y' (UncommunicativeVariableName) (reek)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/649/ : SUCCESS in 18s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1477/ : SUCCESS in 2m 47s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1738/ : FAILURE in 1m 05s

**Comment by kubamahnert** [comment]
https://travis-ci.org/kubamahnert/gooddata-ruby/builds/435551234

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/650/ : SUCCESS in 10s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1478/ : FAILURE in 1m 06s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1739/ : SUCCESS in 2m 39s

**Pull request merged**